File: crates/brainwires-mdap/examples/task_decomposition.rs
//! Task Decomposition Example
//!
//! Demonstrates task decomposition strategies and MDAP cost estimation
//! using the scaling laws from the MAKER paper.
//!
//! Run with: `cargo run -p brainwires-agent --features mdap --example task_decomposition`

use brainwires_mdap::decomposition::{DecomposeContext, SequentialDecomposer, TaskDecomposer};
use brainwires_mdap::scaling::{ModelCosts, estimate_mdap};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    println!("=== Task Decomposition & MDAP Cost Estimation ===\n");

    // 1. Decompose a multi-step task using SequentialDecomposer
    println!("--- 1. Sequential Task Decomposition ---\n");

    let decomposer = SequentialDecomposer::new(5);
    let context = DecomposeContext::new("/home/user/project")
        .with_tools(vec![
            "read_file".to_string(),
            "write_file".to_string(),
            "bash".to_string(),
        ])
        .with_context("Rust web server project");

    let task = "\
1. Read the current handler code in src/handlers.rs
2. Add a new GET /health endpoint that returns 200 OK
3. Register the route in src/router.rs
4. Write a test for the health endpoint in tests/api.rs
5. Run cargo test to verify everything compiles";

    let result = decomposer.decompose(task, &context).await?;

    println!("  Task decomposed into {} subtasks:", result.subtasks.len());
    println!("  Is minimal: {}", result.is_minimal);
    println!("  Total complexity: {:.2}", result.total_complexity);
    println!(
        "  Composition: {}",
        result.composition_function.description()
    );
    println!();

    for subtask in &result.subtasks {
        let deps = if subtask.depends_on.is_empty() {
            "none".to_string()
        } else {
            subtask.depends_on.join(", ")
        };
        println!(
            "  [{}] {} (complexity: {:.2}, depends on: {})",
            subtask.id, subtask.description, subtask.complexity_estimate, deps,
        );
    }

    // 2. Check if simple tasks are considered minimal
    println!("\n--- 2. Minimality Check ---\n");

    let simple_task = "Return the sum of two numbers";
    let complex_task = "1. Parse the input\n2. Validate\n3. Compute\n4. Format output";

    println!(
        "  '{}' is minimal: {}",
        simple_task,
        decomposer.is_minimal(simple_task)
    );
    println!(
        "  Multi-line task is minimal: {}",
        decomposer.is_minimal(complex_task),
    );

    // 3. MDAP cost estimation with estimate_mdap
    println!("\n--- 3. MDAP Cost Estimation ---\n");

    let scenarios = vec![
        ("Simple (5 steps, p=0.95)", 5, 0.95, 0.90, 0.003, 0.95),
        ("Moderate (10 steps, p=0.85)", 10, 0.85, 0.85, 0.003, 0.95),
        ("Complex (20 steps, p=0.75)", 20, 0.75, 0.80, 0.003, 0.95),
        (
            "High-reliability (10 steps, p=0.90, t=0.99)",
            10,
            0.90,
            0.90,
            0.003,
            0.99,
        ),
    ];

    println!(
        "  {:<45} {:>5} {:>8} {:>10} {:>8}",
        "Scenario", "k", "Calls", "Cost ($)", "P(success)",
    );
    println!("  {}", "-".repeat(80));

    for (name, steps, p, v, cost, target) in &scenarios {
        let estimate = estimate_mdap(*steps, *p, *v, *cost, *target)?;
        println!(
            "  {:<45} {:>5} {:>8} {:>10.4} {:>7.1}%",
            name,
            estimate.recommended_k,
            estimate.expected_api_calls,
            estimate.expected_cost_usd,
            estimate.success_probability * 100.0,
        );
    }

    // 4. Compare model costs
    println!("\n--- 4. Model Cost Comparison ---\n");

    let models: Vec<(&str, ModelCosts)> = vec![
        ("Claude Sonnet", ModelCosts::claude_sonnet()),
        ("Claude Haiku", ModelCosts::claude_haiku()),
        ("GPT-4o", ModelCosts::gpt4o()),
        ("GPT-4o Mini", ModelCosts::gpt4o_mini()),
    ];

    let input_tokens = 500;
    let output_tokens = 200;

    println!(
        "  {:<16} {:>12} {:>12} {:>14}",
        "Model", "Input/1K", "Output/1K", "Per Call Cost",
    );
    println!("  {}", "-".repeat(58));

    for (name, costs) in &models {
        let call_cost = costs.estimate_call_cost(input_tokens, output_tokens);
        println!(
            "  {:<16} {:>11.5}$ {:>11.5}$ {:>13.6}$",
            name, costs.input_per_1k, costs.output_per_1k, call_cost,
        );
    }

    // 5. Full cost projection: model costs * MDAP scaling
    println!("\n--- 5. Full MDAP Cost Projection (10 steps, p=0.85, target=0.95) ---\n");

    let steps = 10_u64;
    let p = 0.85;
    let v = 0.90;
    let target = 0.95;

    println!(
        "  {:<16} {:>14} {:>10} {:>12}",
        "Model", "Per Call ($)", "Est. Calls", "Total ($)",
    );
    println!("  {}", "-".repeat(56));

    for (name, costs) in &models {
        let call_cost = costs.estimate_call_cost(input_tokens, output_tokens);
        let estimate = estimate_mdap(steps, p, v, call_cost, target)?;
        println!(
            "  {:<16} {:>14.6} {:>10} {:>12.4}",
            name, call_cost, estimate.expected_api_calls, estimate.expected_cost_usd,
        );
    }

    println!("\n=== Done ===");
    Ok(())
}

File: deno/examples/agents/task_decomposition.ts
// Example: Task Decomposition & MDAP Cost Estimation
// Demonstrates task decomposition strategies and MDAP cost estimation
// using the scaling laws from the MAKER paper.
// Run: deno run deno/examples/agents/task_decomposition.ts

import {
  atomicDecomposition,
  compositeDecomposition,
  createAtomicSubtask,
  defaultDecomposeContext,
  estimateCallCost,
  estimateMdap,
  type MdapSubtask,
  MODEL_COSTS,
  type ModelCosts,
} from "@rullama/agent";

async function main() {
  console.log("=== Task Decomposition & MDAP Cost Estimation ===\n");

  // 1. Build a decomposition context and create subtasks manually
  console.log("--- 1. Sequential Task Decomposition ---\n");

  const _context = {
    ...defaultDecomposeContext("/home/user/project"),
    availableTools: ["read_file", "write_file", "bash"],
    additionalContext: "Rust web server project",
  };

  // Create subtasks with full control over fields
  const subtasks: MdapSubtask[] = [
    {
      id: "step-1",
      description: "Read the current handler code in src/handlers.rs",
      inputState: null,
      dependsOn: [],
      complexityEstimate: 0.2,
    },
    {
      id: "step-2",
      description: "Add a new GET /health endpoint that returns 200 OK",
      inputState: null,
      dependsOn: ["step-1"],
      complexityEstimate: 0.4,
    },
    {
      id: "step-3",
      description: "Register the route in src/router.rs",
      inputState: null,
      dependsOn: ["step-2"],
      complexityEstimate: 0.3,
    },
    {
      id: "step-4",
      description: "Write a test for the health endpoint in tests/api.rs",
      inputState: null,
      dependsOn: ["step-3"],
      complexityEstimate: 0.5,
    },
    {
      id: "step-5",
      description: "Run cargo test to verify everything compiles",
      inputState: null,
      dependsOn: ["step-4"],
      complexityEstimate: 0.3,
    },
  ];

  const result = compositeDecomposition(subtasks, { kind: "sequence" });

  console.log(`  Task decomposed into ${result.subtasks.length} subtasks:`);
  console.log(`  Is minimal: ${result.isMinimal}`);
  console.log(`  Total complexity: ${result.totalComplexity.toFixed(2)}`);
  console.log(`  Composition: ${result.compositionFunction.kind}`);
  console.log();

  for (const subtask of result.subtasks) {
    const deps = subtask.dependsOn.length === 0
      ? "none"
      : subtask.dependsOn.join(", ");
    console.log(
      `  [${subtask.id}] ${subtask.description} (complexity: ${
        subtask.complexityEstimate.toFixed(2)
      }, depends on: ${deps})`,
    );
  }

  // 2. Check if simple tasks are considered minimal (atomic)
  console.log("\n--- 2. Minimality Check ---\n");

  const simpleSubtask = createAtomicSubtask("Return the sum of two numbers");
  const atomicResult = atomicDecomposition(simpleSubtask);
  console.log(
    `  'Return the sum of two numbers' is minimal: ${atomicResult.isMinimal}`,
  );
  console.log(
    `  Multi-step task is minimal: ${result.isMinimal}`,
  );

  // 3. MDAP cost estimation with estimateMdap
  console.log("\n--- 3. MDAP Cost Estimation ---\n");

  const scenarios: Array<[string, number, number, number, number, number]> = [
    ["Simple (5 steps, p=0.95)", 5, 0.95, 0.90, 0.003, 0.95],
    ["Moderate (10 steps, p=0.85)", 10, 0.85, 0.85, 0.003, 0.95],
    ["Complex (20 steps, p=0.75)", 20, 0.75, 0.80, 0.003, 0.95],
    [
      "High-reliability (10 steps, p=0.90, t=0.99)",
      10,
      0.90,
      0.90,
      0.003,
      0.99,
    ],
  ];

  console.log(
    `  ${"Scenario".padEnd(45)} ${"k".padStart(5)} ${"Calls".padStart(8)} ${
      "Cost ($)".padStart(10)
    } ${"P(success)".padStart(8)}`,
  );
  console.log(`  ${"-".repeat(80)}`);

  for (const [name, steps, p, v, cost, target] of scenarios) {
    const estimate = estimateMdap(steps, p, v, cost, target);
    console.log(
      `  ${name.padEnd(45)} ${String(estimate.recommendedK).padStart(5)} ${
        String(estimate.expectedApiCalls).padStart(8)
      } ${estimate.expectedCostUsd.toFixed(4).padStart(10)} ${
        (estimate.successProbability * 100).toFixed(1).padStart(7)
      }%`,
    );
  }

  // 4. Compare model costs
  console.log("\n--- 4. Model Cost Comparison ---\n");

  const models: Array<[string, ModelCosts]> = [
    ["Claude Sonnet", MODEL_COSTS.claudeSonnet],
    ["Claude Haiku", MODEL_COSTS.claudeHaiku],
    ["GPT-4o", MODEL_COSTS.gpt4o],
    ["GPT-4o Mini", MODEL_COSTS.gpt4oMini],
  ];

  const inputTokens = 500;
  const outputTokens = 200;

  console.log(
    `  ${"Model".padEnd(16)} ${"Input/1K".padStart(12)} ${
      "Output/1K".padStart(12)
    } ${"Per Call Cost".padStart(14)}`,
  );
  console.log(`  ${"-".repeat(58)}`);

  for (const [name, costs] of models) {
    const callCost = estimateCallCost(costs, inputTokens, outputTokens);
    console.log(
      `  ${name.padEnd(16)} ${costs.inputPer1k.toFixed(5).padStart(11)}$ ${
        costs.outputPer1k.toFixed(5).padStart(11)
      }$ ${callCost.toFixed(6).padStart(13)}$`,
    );
  }

  // 5. Full cost projection: model costs * MDAP scaling
  console.log(
    "\n--- 5. Full MDAP Cost Projection (10 steps, p=0.85, target=0.95) ---\n",
  );

  const projSteps = 10;
  const projP = 0.85;
  const projV = 0.90;
  const projTarget = 0.95;

  console.log(
    `  ${"Model".padEnd(16)} ${"Per Call ($)".padStart(14)} ${
      "Est. Calls".padStart(10)
    } ${"Total ($)".padStart(12)}`,
  );
  console.log(`  ${"-".repeat(56)}`);

  for (const [name, costs] of models) {
    const callCost = estimateCallCost(costs, inputTokens, outputTokens);
    const estimate = estimateMdap(
      projSteps,
      projP,
      projV,
      callCost,
      projTarget,
    );
    console.log(
      `  ${name.padEnd(16)} ${callCost.toFixed(6).padStart(14)} ${
        String(estimate.expectedApiCalls).padStart(10)
      } ${estimate.expectedCostUsd.toFixed(4).padStart(12)}`,
    );
  }

  console.log("\n=== Done ===");
}

await main();
